### PR TITLE
feat: wildcards

### PR DESCRIPTION
**Pull request by brprice**

### CODE REVIEW

**Comment by brprice** [comment]
status: needs to be bumped to a backend pin from main, when https://github.com/hackworthltd/primer/pull/1049 has been merged. The second commit is simply things to consider in review

**Comment by brprice** [comment]
There is some bikeshedding possible on definitions of colors etc. We currently have pattern wildcards are always green, matching AST patterns & constructors but primitive constructors & patterns are black. This looks somewhat odd when doing a case on a primitive.

**Comment by brprice** [comment]
There is some weird UI staleness that I don't understand:
- create a case with just the wildcard branch, and select the `match` node
- note there is a "add a branch" action offered but not a "delete branch", as expected
- add a branch
- note that the actions offered have not updated: there is no "delete a branch" offered, although there should be

@georgefst any ideas?

**Comment by brprice** [comment]
I also seem to be having some problems with local development (which I suspect are unrelated to this PR, but have not yet confirmed):
- sometimes when entering a shell I get an error about mismatched pnpm versions
- sometimes when running `pnpm watch`, `eslint` complains about parsing errors on `primer.sqlite3` and `README.md`

File: pnpm-lock.yaml [hunk 1]
```diff
@@ -1,4 +1,8 @@
-lockfileVersion: '6.0'
+lockfileVersion: '6.1'
+
+settings:
+  autoInstallPeers: true
+  excludeLinksFromLockfile: false
 
 overrides:
```
Comment (brprice) [inline pnpm-lock.yaml:7]: The changes to pnpm-lock.yaml really should not be in this PR. They probably are the reason that I am having issues with the shell

**Comment by dhess** [comment]
My comments on how this feature currently works. This isn't a review as such; I just want to get my thoughts down so we can return to these comments and ideas for later work, and maybe hash some of these ideas out here in the PR discussion.

### `match` action now defaults to only adding the wildcard

(For posterity, before this PR, the `match` action would automatically create a pattern for each constructor in the subject's type, for non-primitive types, anyway.)

For primitives, this behavior makes sense as the default, for obvious reasons.

For non-primitives, I'm not sure. Perhaps the default behavior should depend on the level: in beginner mode, wildcard patterns may be more confusing than helpful and I wonder if we should stick to our "explicit is better than implicit" design principle.

At intermediate and expert levels, I think it would make sense to offer 2 different `match` actions: one that splats out all explicit constructors, and one that creates only the wildcard pattern.

### The output value of a newly-added branch

It appears that if one of the existing branches is the wildcard pattern, then when adding a new branch, the new branch's output will be automatically assigned to whatever the output value of the wildcard pattern is.

I'm of two minds on this. On the one hand, it's a nice touch, and it makes sense that if *any* output value is automatically assigned, it should be the wildcard's output value, because this preserves the behavior of the `match` expression. But on the other hand:

* There's a case to be made that the default output value (if any) should *not* be the same as the wildcard output value, since presumably the whole point of creating a new branch case is to distinguish it from the default (wildcard) case.
* I wonder if beginners, at least, might be confused by the fact that the new branch has any output value at all other than hole (`?`). Would they jump to the conclusion that it can't be changed, because it was pre-populated? 
* Would it lead to more program errors? Students might not notice that the automatically-provided output value is not what they want.


### The wildcard pattern visual

Personally I really like the visual style of the wildcard pattern as it's implemented here, modulo some tweaking and perhaps a different choice of wildcard symbol. I think @georgefst might be less sold on it, though. My concern with using an empty pattern var box is that students may see it and think their program is incomplete. I like the idea of using the pattern var box, but I think we would need to fill it with some symbol, and then the issue is, what symbol to use? We don't want to pick a Unicode character that could also be used for a pattern variable name, but we also don't want to reserve any of those characters as special, so it seems like a catch-22.

### The add/remove UX

I think that offering the add/remove branch options when the `match` node is selected makes sense, but it does seem to lack some discoverability. (It wasn't obvious to me on first use how to add/remove branches.) Then again, it should be easy to remember and get used to once you've discovered it, and it does seem like a natural action to offer for selected `match` nodes.

It's obviously out of scope for this PR, but I think we could do more. For example, you should be able to select a pattern box and be offered actions either to a) delete it or b) convert it to (or merge it into an existing) the wildcard pattern. It might also make sense to offer the "add a branch" whenever any pattern box is selected, though that might not work well once we have support for "deep" patterns.

This is probably also a case where adding hotkey support for add (e.g., `+`) and delete (e.g., `del`) would be helpful.

**Review by dhess — APPROVED** [review]
Thanks, I'm impressed by how quickly you got this into shape!

Besides the deeper questions I raised in my comment in the discussion on this PR, there are a few things I would like to bikeshed/change in this implementation, such as how to style the wildcard pattern, and the need for some kind of text caption when the student is prompted to provide a primitive constructor for a new branch, but in the interest of our upcoming deadline, let's get this committed first, and then we can revisit these couple of points as time allows. The important thing is that we now have working wildcard patterns!

**Comment by dhess** [comment]
Closing in favor of #957. I'll copy my comments from this thread there.

**Pull request closed without merging**